salt/states/archive.py

```diff
@@ -1387,6 +1443,68 @@ def extracted(name,
             ret['comment'] = exc.strerror
             return ret
 
+    if skip_same_file:
+        extracted_files = []
+        log.debug('Extracting %s to %s', name, origin_name)
+        mng_files, mng_dirs, mng_links = _file_lists(name, origin_name)
+        for dest_dir, src_dir in mng_dirs:
+            try:
+                if not os.path.isdir(dest_dir):
+                    if os.path.exists(dest_dir):
+                        __salt__['file.remove'](dest_dir)
+                    stats = __salt__['file.stats'](src_dir)
+                    __states__['file.directory'](dest_dir, user=stats.get('user'), group=stats.get('group'), file_mode=stats.get('mode'))
+            except Exception as exc:
+                __salt__['file.remove'](name)
+                ret['changes'] = {}
+                log.debug(traceback.format_exc())
+                return _error(ret, 'Unable to manage dir: {0}'.format(exc))
+
+        for dest_file, src_file in mng_files:
+            try:
+                stats = __salt__['file.stats'](src_file)
+                ret = __salt__['file.manage_file'](
```

Comment:
This will override the `ret` dict for the state with the results of `file.manage_file`. This is a _bad_ idea.

Reply:
In this scene, the main task is extracting archives, other tasks like manage_file dont need return too much information. Maybe fast failure is recommended.

Reply:
This still hasn't been changed, and it _needs_ to be changed. Since `ret` holds the state's return dictionary, replacing it with the result of a remote execution function will cause errors in the state compiler if we then attempt to do a `return ret`.

Reply:
I also don't understand why this is using `file.manage_file`, which hasn't been explained since I asked a few weeks ago. This is a _lot_ of extra overhead when all we care about is whether or not the file is different. Simply comparing the checksums and replacing only when the checksum differs, should be sufficient. Perhaps an extra check to see if the paths are links.